lib/src/dense_set.rs

```diff
@@ -0,0 +1,1282 @@
+use crate::data_structures::{Reg, RegClass, Set};
+use std::{cmp, fmt};
+
+//=============================================================================
+// RegSet
+
+const BLOCK_SIZE: usize = 64;
+const NUMBER_REAL_REG: usize = 64;
+
+pub struct BitSet {
+    data: Vec<u64>,
+    reg_vec: Vec<Reg>,
+}
+
+pub trait RegSet {
+    fn empty() -> Self;
+    /// Return a new BitSet with one item set.
+    fn unit(item: Reg) -> Self;
+    /// Return a new BitSet with two items set.
+    fn two(item1: Reg, item2: Reg) -> Self;
+    fn card(&self) -> usize;
+    /// Set the item to 1.
+    fn insert(&mut self, item: Reg);
+    fn data_insert(&mut self, item: Reg);
+    fn vec_insert(&mut self, item: Reg);
+    /// If the item is set, clear the item to 0.
+    /// Do nothing otherwise.
+    fn delete(&mut self, item: Reg);
+    fn vec_delete(&mut self, item: Reg);
+    fn data_delete(&mut self, item: Reg);
+    fn vec_get_item(&self, index: usize, real: bool) -> Option<&Reg>;
+    fn is_empty(&self) -> bool;
+    /// Toggle the item.
+    fn contains(&self, item: Reg) -> bool;
+    fn intersect(&mut self, other: &Self);
+    fn union(&mut self, other: &Self);
+    fn remove(&mut self, other: &Self);
+    fn intersects(&self, other: &Self) -> bool;
+    fn is_subset_of(&self, other: &Self) -> bool;
+    fn equals(&self, other: &Self) -> bool;
+    fn from_vec(vec: Vec<Reg>) -> Self;
+    fn to_vec(&self) -> Vec<Reg>;
+}
+
+fn get_reg_index(item: Reg) -> usize {
+    if item.is_real() {
+        item.get_index()
+    } else {
+        NUMBER_REAL_REG + item.get_index()
+    }
+}
+
+fn get_offset(reg_index: usize) -> usize {
+    reg_index % BLOCK_SIZE
+}
+
+fn get_data_index(reg_index: usize) -> usize {
+    reg_index / BLOCK_SIZE
+}
+
+impl RegSet for BitSet {
+    fn empty() -> Self {
+        let mut d = Vec::<u64>::new();
+        d.push(0);
```

Comment:
There's no need to push an empty entry here, I think this has been done to accommodate the needs of the iterator. Instead, can you change the iterator so `set_iter` is an `Option` instead, set to `None` in initially, and so the iterator works correctly in this case?

Reply:
Hello,
Thanks for the review! It's done f1961fb793b2e9d4a7c94ee388a43902571e0a49.